### PR TITLE
qt: Hide non PKHash-Addresses in signing address book

### PR DESCRIPTION
**Pull request by cvengler**
[Video Demo](https://www.youtube.com/watch?v=T-Rp2pFRmzY)

This PR hides all non PKHash addresses in the signing GUI in the Address Book when it is opened through the signing dialog, as non PKHash addresses are useless there.

### CODE REVIEW

**Comment by ysangkok** [comment]
`PKHashOnly` is a variable but is not snake-case as it should be according to [dev notes](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#coding-style-c).

**Comment by theStack** [comment]
Is it really good UI design to just hide addresses? From a user perspective I'd rather prefer to see a message that explicitely tells me that bech32 address signing is not supported yet in Bitcoin Core, rather than being confused on why the addresses don't even show up in the Sign/Verify dialog in the first place.

**Comment by emilengler** [comment]
> Is it really good UI design to just hide addresses?

Yes, it just shows what really matters

**Comment by emilengler** [comment]
@ysangkok Fixed

**Comment by MarcoFalke** [comment]
It should explicitly  say in the window that non-pkh addresses are hidden?

**Comment by emilengler** [comment]
> It should explicitly say in the window that non-pkh addresses are hidden?

Good idea. I just meant that the current approach isn't that good. Maybe I understood theStack wrong.

File: src/qt/addresstablemodel.cpp [hunk 1]
```diff
@@ -75,12 +75,14 @@ class AddressTablePriv
...
         parent(_parent) {}
 
-    void refreshAddressTable(interfaces::Wallet& wallet)
+    void refreshAddressTable(interfaces::Wallet& wallet, bool pk_hash_only = false)
     {
         cachedAddressTable.clear();
         {
             for (const auto& address : wallet.getAddresses())
             {
+                if (pk_hash_only && address.dest.type() != typeid(PKHash))
```
Comment (Empact) [inline src/qt/addresstablemodel.cpp]: `#include <typeinfo>`
https://en.cppreference.com/w/cpp/language/typeid

File: src/qt/signverifymessagedialog.cpp [hunk 2]
```diff
@@ -89,6 +89,7 @@ void SignVerifyMessageDialog::on_addressBookButton_SM_clicked()
 {
     if (model && model->getAddressTableModel())
     {
+        model->refresh(true);
```
Comment (Empact) [inline src/qt/signverifymessagedialog.cpp]: `model->refresh(true /* pk_hash_only */);`

**Comment by Empact** [comment]
I'm inclined to agree with @theStack that showing disabled addresses would be more clear than a message at the top, but I could go either way.

Concept ACK

**Comment by emilengler** [comment]
> I'm inclined to agree with @theStack that showing disabled addresses would be more clear than a message at the top, but I could go either way.

There is a message for it, see `addressbookpage.cpp:109`

File: src/qt/signverifymessagedialog.cpp (hunk 2 above)
Comment (emilengler) [reply src/qt/signverifymessagedialog.cpp]: Implemented with your other suggestion in c4ea501e96363e937200bc97b8e2d78162bdb699

File: src/qt/addressbookpage.cpp [hunk 3]
```diff
@@ -106,7 +106,7 @@ AddressBookPage::AddressBookPage(const PlatformStyle *platformStyle, Mode _mode,
         ui->newAddress->setVisible(true);
         break;
     case ReceivingTab:
-        ui->labelExplanation->setText(tr("These are your Bitcoin addresses for receiving payments. Use the 'Create new receiving address' button in the receive tab to create new addresses."));
+        ui->labelExplanation->setText(tr("These are your Bitcoin addresses for receiving payments. Use the 'Create new receiving address' button in the receive tab to create new addresses.\nSigning is only possible with addresses of the type 'legacy'."));
```
Comment (luke-jr) [inline src/qt/addressbookpage.cpp]: Suggest breaking this into two `tr` calls so we 1) don't lose existing translations of the first part, and 2) can remove or change the second part without resetting translations every time.

File: src/qt/walletmodel.cpp [hunk 4]
```diff
@@ -563,3 +563,8 @@ bool WalletModel::isMultiwallet()
 {
     return m_node.getWallets().size() > 1;
 }
+
+void WalletModel::refresh(bool pk_hash_only)
+{
+    addressTableModel = new AddressTableModel(this, pk_hash_only);
```
Comment (luke-jr) [inline src/qt/walletmodel.cpp]: I'm not sure it makes sense to use the same cache with context-dependent factors like this.

Probably would be best to keep the same AddressTableModel with all addresses, and then a FilteredAddressTableModel(AddressTableModel*) to hide them...

File: src/qt/addressbookpage.cpp (hunk 3 above)
Comment (emilengler) [reply src/qt/addressbookpage.cpp]: ACK

**Comment by jonasschnelli** [comment]
Code Review ACK c4ea501e96363e937200bc97b8e2d78162bdb699

**Pull request merged**